src/interpreter/executors/instruction/tests/reserved_registers.rs

```diff
@@ -0,0 +1,140 @@
+use super::*;
+use quickcheck::TestResult;
+use quickcheck_macros::quickcheck;
+
+// Ensure none of the opcodes can write to reserved registers
+#[quickcheck]
+fn cant_write_to_reserved_registers(opcode: u8, fuzz: u16) -> TestResult {
+    // opcode || REG_ONE || random fuzz
+    let attempt_reserved_reg_access = ((opcode as u32) << 24) | (1 << 18) | ((fuzz >> 4) as u32);
+    let instruction = Instruction::new(attempt_reserved_reg_access);
+    let opcode = Opcode::new(instruction);
+    // skip undefined opcodes
+    if matches!(opcode, Opcode::Undefined) {
+        return TestResult::discard();
+    }
+
+    let mut vm = Interpreter::with_memory_storage();
+
+    vm.init_script(CheckedTransaction::default())
+        .expect("Failed to init VM");
+    let res = vm.instruction(instruction);
+
+    if writes_to_ra(opcode) {
+        // if this opcode writes to $rA, expect an error since we're attempting to use a reserved register
+        if !matches!(
+            res,
+            Err(InterpreterError::PanicInstruction(r)) if r.reason() == &ReservedRegisterNotWritable
+        ) {
+            return TestResult::error(format!(
+                "expected ReservedRegisterNotWritable error {:?}",
+                (opcode, &res)
+            ));
+        }
+    } else {
+        // Ensure REG_ONE wasn't changed.
+        // While not a perfect guarantee against the opcode writing a value
+        // to an invalid register, this increases the likelihood of detecting
+        // erroneous register access.
+        if vm.registers[REG_ONE] != 1 {
+            return TestResult::error("reserved register was modified!");
+        }
+
+        // throw err if a ReservedRegisterNotWritable err was detected outside our writes_to_ra check
+        // This would likely happen if the opcode wasn't properly marked as true in `writes_to_ra`
+        if matches!(
+            res,
+            Err(InterpreterError::PanicInstruction(r)) if r.reason() == &ReservedRegisterNotWritable
+        ) {
+            return TestResult::error(format!(
+                "unexpected ReservedRegisterNotWritable, test configuration may be faulty {:?}",
+                (opcode, &res)
+            ));
+        }
+    }
+    TestResult::passed()
+}
+
+// determines whether a given opcode stores a value into $rA
+fn writes_to_ra(opcode: Opcode) -> bool {
```

Comment:
It might be a good idea to instead expose this property from `fuel-asm` itself as `Opcode::writes_to_ra()`. Although I cannot imagine it being used for anything else than this test, I think it might help us to track breaking changes in there. Or is the intent that `fuel-asm` only concerns itself with representation, leaving the semantics to `fuel-vm`?

Reply:
The advantage of moving this to Opcode is that the VM could generalize the checks for reserved register access across any instruction. The disadvantage is that we end up with extra branching hits for opcodes that don't need the reserved register access check.

I generally prefer to keep utilities like this close to where they are used, until we see a clear shared need across the board. Since fuel-vm is entirely responsible for deciding which opcodes need writable register checks, I think it makes sense to keep this util in the VM for now. There are likely a lot of other shared validation rules that apply to specific subcategories of opcodes, but since they are enforced by fuel-vm it may be easier to manage it like this for now.